Chapter 002 remove docker troubleshooting (#24)
psbanka/build-your-own-internet: chapters/002-smol-internet/README.md

^C
4 packets captured
4 packets received by filter
0 packets dropped by kernel
```

We can see why `tara` is able to respond to `boudi` from the first line of the (non-default) output:

```bash
18:50:57.464168 IP 10.1.2.3 > 10.1.2.2: ICMP echo request, id 106, seq 1, length 64
```

Here we see the incoming `ping` or `ICMP echo request`. The source machine is `10.1.2.3`, or, if you remember what we added to our [docker-compose.yml file](./docker-compose.yml), `boudi`'s IP address on `doggonet`! This means that the request is coming from an IP address on a network that `tara` has an interface on! `tara` can respond! The next line is `tara`, `10.1.2.2` responding back to `boudi`, `10.1.2.3`, with an `ICMP echo reply`.

> **What's with those ARP requests?**

YOu may or may not see in your own session some odd looking packets identified as `ARP` in the tcpdump:

```bash
18:53:54.166970 ARP, Request who-has 10.1.2.3 tell 10.1.2.2, length 28
18:53:54.167080 ARP, Request who-has 10.1.2.2 tell 10.1.2.3, length 28
18:53:54.167123 ARP, Reply 10.1.2.2 is-at 02:42:0a:01:02:02, length 28
18:53:54.167168 ARP, Reply 10.1.2.3 is-at 02:42:0a:01:02:03, length 28
```

> We go over this in more detail in [ip-and-mac-addresses.md in the appendix](../appendix/ip-and-mac-addresses.md), but let's look at a high level at what's going on here. IP addresses, like `10.1.2.3`, are used by machines for identifying where packets should be routed across an internet. So what we've been working with so far is designed to help machines communicate when there are multiple networks. HOWEVER. Within a network, machines are not identified by an IP address, but instead by a MAC address. In order for packets to be forwarded from one machine on a network to another machine on the same network, each machine needs to discover the MAC address that corresponds to the IP address identified in the packets. ARP, or Address Resolution Protocol, is the process by which this is done.

Reviewer: ```suggestion
We go over this in more detail in [ip-and-mac-addresses.md in the appendix](../appendix/ip-and-mac-addresses.md), but let's look at a high level at what's going on here. IP addresses, like `10.1.2.3`, are used by machines for identifying where packets should be routed across an internet. So what we've been working with so far is designed to help machines communicate when there are multiple networks. HOWEVER. Within a network, machines are not identified by an IP address, but instead by a MAC address. In order for packets to be sent from one machine on a network to another machine on the same network, each machine needs to discover the MAC address that corresponds to the IP address identified in the packets. ARP, or Address Resolution Protocol, is the process by which this is done.
```
I get a heebie-jeebie when I see the term `forwarded` here, since it's routers that forward packets (okay, switches sort of do this too, but we don't really talk about that as "forwarding"). In the layer-2 case, we are really just putting the packet on the wire and having the other machine pick it up directly.

Author: Heh. I remember having a similar thought when I wrote this. I made a mental note to come back and think on what word to use here, but clearly.... I forgot 😳

> Let's read what's happening with the `ARP` requests we see above:

Reviewer: Should we have a central place where we talk about the commands we use and the flags we use with them?

Author: I like this idea. I think we should leave the more detailed explanations in the chapter, but then we can have a reference guide for people to use for findability...

I was also thinking that at some point I wanted to include links to [specific comix from Julia Evans](https://wizardzines.com/comics/) in our explanations. For example:
* [ping & traceroute](https://wizardzines.com/comics/ping/)
* [tcpdump](https://wizardzines.com/comics/tcpdump/)
* maybe something on [containers](https://wizardzines.com/comics/why-containers/)
* we reference [iptables](https://wizardzines.com/comics/iptables/) in chapter 003

Maybe we can add a `## Helpful Links` section to our chapters? Or actually do the work to figure out where the best place to link these things is 😆

Author: @tundal45  and I talked about this and i'll make another PR that includes a reference guide with a first stab and linking to it where useful.

```bash
18:53:54.167123 ARP, Reply 10.1.2.2 is-at 02:42:0a:01:02:02, length 28
```

Here's the MAC address for `10.1.2.2`, so now the machine on `10.1.2.3` knows who to respond to, which it does in the next `Reply`:

```bash
18:53:54.167168 ARP, Reply 10.1.2.3 is-at 02:42:0a:01:02:03, length 28
```

Now back to our regularly scheduled exploration!

### Can `tara` ping `boudi` on the `squasheeba` network?

We've already seen that `tara` can respond to `boudi`'s `ping`s that were issued from `boudi`'s interface on `doggonet`. We can double check that `tara` can initiate the `ping`, just for fun:

```bash
root@tara:/# ping 10.1.2.3 -c 2
PING 10.1.2.3 (10.1.2.3) 56(84) bytes of data.
64 bytes from 10.1.2.3: icmp_seq=1 ttl=64 time=0.245 ms
64 bytes from 10.1.2.3: icmp_seq=2 ttl=64 time=0.242 ms

--- 10.1.2.3 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1041ms
rtt min/avg/max/mdev = 0.242/0.243/0.245/0.001 ms
```

Now... let's see if `tara` is able to ping `boudi`'s interface on the `squasheeba` network...

```bash
root@tara:/# ping 10.1.1.3
ping: connect: Network is unreachable
```

> Lovely! This is because, while `boudi` has an interface on the `doggonet` network and `tara` and `boudi` can directly communicate on that network, `tara` doesn't know anything about the `squasheeba` network; not even that `boudi` knows about it. We need to define a route for `tara` that tells it how to get to the `squasheeba` network!

Reviewer: GENERAL QUESTION: should we use the names of these networks, (e.g. `squasheeba`), or should we use network addresses (e.g. `10.1.1.0/24`) ? 

I am starting to feel like it's a little weird to name networks, but it is also maybe a "pets versus cattle" thing and we tend to abandon it later. Perhaps it makes the introduction feel more warm having fun names for things. Thoughts?

Author: I'm inclined not to abandon the names. 2 reasons:
* The network addresses are similar enough that using names makes it easier to see which network we're talking about. And we could, absolutely, make the addresses more distinct from each other, but I think the friendly names just makes it more approachable?
* if we abandon the network names, then we should look at abandoning the machine names for consistency. Then we're in a whole other realm of confusion about which machine is which. The approachable names lets us start teaching concepts without making things harder than they need to be.

Author: After thinking about this overnight, I do think we should pick a more accessible name than `squasheeba`... Since we have `doggonet`, how do you feel about `caternet` as the other network?

Author: @tundal45 and I chatted about this. We'd prefer to keep referencing the names, but we will update `squasheeba` => `caternet` for readability by broader (non-Fastly) audiences.

Author: https://github.com/psbanka/build-your-own-internet/pull/25

### Make `tara` ping hosts on the `squasheeba` network

> The first thing we need to do is add a route from `tara` to the `squasheeba` network via `boudi`. Because `boudi` has routes to both `doggonet` and `squasheeba`, `boudi` can act as the gateway between the two. We can manage routes on our machines using the `ip route` command:

Reviewer: ```suggestion
The first thing we need to do is add a route from `tara` to the `squasheeba` network via `boudi`. Because `boudi` has routes to both `doggonet` and `squasheeba`, `boudi` can act as the router between the two. We can manage routes on our machines using the `ip route` command:
```
I think we're preferring the term `router` instead of the synonym `gateway`

```bash
root@tara:/# ip route add 10.1.1.0/24 via 10.1.2.3
```

> This command defines the network, `10.1.1.0/24` and then says that routes to that network should use a machine that exists on a network it has an interface on, namely, `boudi` at `10.1.2.3`. Now, if we check the routes that `tara` knows about, we'll see the route defined in `tara`'s routing table:

Reviewer: ```suggestion
This command identifies the network, `10.1.1.0/24` (a.k.a. "squasheeba") and then says: "Any time you have a packet for this network, you should send it to `10.1.2.3` (a.k.a. `boudi`), cuz that dude knows all about it." 

Now, if we check the routes that `tara` knows about, we'll see the route defined in `tara`'s routing table:
```

`and then says that routes to that network should use a machine that exists on a network it has an interface on` felt a little awkward to me

Author: YES! Thank you! I had the same thought! Brilliant cleanup here!

  * checks the destination IP
  * sees that the request is not for itself
  * checks whether or not to forward the packet

>   * finds out that it should route the packets - Check the end of this chapter for [how does a machine know if it should forward packets](#how-does-a-machine-know-if-it-should-forward-packets)

Reviewer: ❤️ such improve ❤️

### Tell `pippin` how to respond to `tara`

> We've already seen this in action. At this point, we need to tell `pippin` how to find the `doggonet` network. We did this earlier in teaching `tara` how to find the `squasheeba` network. We're going to leave this as an exercise for the reader to attempt on their own. If you need some guidance, review the [Make `tara` ping hosts on the squasheeba network](#make-tara-ping-hosts-on-the-squasheeba-network) section.

Reviewer: 👍 It strikes me that we should maybe write a markdown linter/tester that verifies that all our links in this document go to an actual file

Author: I think the markdown linter I'm already using does that? At least, I remember seeing a yellow squiggle under a broken link. But maybe I'm misremembering that?

      - net.ipv4.ip_forward=0
```

> For the sake of ensuring the rest of this chapter works as expected, we will not disable ip forwarding. HOWEVER, going forward, we are going to disable ip forwarding for any machine we do not explicitly want to be a router.

Reviewer: 💡
